### DISCUSSION TITLE
Psalm annotation in class docblock is ignored

### DISCUSSION THREAD

**Discussion by rizqyhi** [vimeo/psalm#9056] [Q&A] [answered]
Hi all, I have a problem where the psalm annotation is being ignored. We're using [`myclabs/php-enum`](https://github.com/myclabs/php-enum) which internally they're marking their class with `@psalm-immutable`. It trigger this error when I run psalm check:

> ERROR: MissingImmutableAnnotation - MembershipStatus.php:19:38 - MyCLabs\Enum\Enum is marked @psalm-immutable, but MembershipStatus is not marked @psalm-immutable (see https://psalm.dev/213)

I tried to add `@psalm-immutable` to my class' docblock, like so:

```
/**
 * ...
 * @method static MembershipStatus UNVERIFIED_USER
 * @method static MembershipStatus VERIFIED_USER
 * @psalm-immutable
 */
```

But psalm still report the same error as above. I also tried to run pslam with `--no-cache` option and still no luck. Is there anything I missed?

Thanks for the help.

**Comment by orklah** [comment, marked as answer]
Do you use attributes on this class?

Could you try moving that specific annotation at the top? It could be a parser issue (also if you could post the whole docblock, it would help)

**Reply by rizqyhi** [reply]
Hi @orklah, thank you for your reply. I don't really get what do you mean by attributes, but here is the whole class:

```php
namespace Domain\User\Enums;

use MyCLabs\Enum\Enum;

/**
 * Class MembershipStatus
 * @package Domain\Membership\Enums
 * @method static MembershipStatus UNVERIFIED_USER
 * @method static MembershipStatus VERIFIED_USER
 */
class MembershipStatus extends Enum
{
    private const UNVERIFIED_USER = 0;
    private const VERIFIED_USER = 1;
}
```

Yes, I've tried moving the annotation above all and still not working. The only way it could work is by placing the annotation in separate docblock, so like:

```php
/**
 * Class MembershipStatus
 * @package Domain\Membership\Enums
 * @method static MembershipStatus UNVERIFIED_USER
 * @method static MembershipStatus VERIFIED_USER
 */
/** @psalm-immutable */
class MembershipStatus extends Enum
{
    private const UNVERIFIED_USER = 0;
    private const VERIFIED_USER = 1;
}
```

We try to avoid this approach because we think its kinda cluttering the code.

**Reply by orklah** [reply]
This looks like a bug, the annotation should behave the same whether or not it's in the common docblock. Can you try putting it before the @package one?

**Reply by rizqyhi** [reply]
I've tried putting it anywhere in the common docblock and still no luck. But I found something, when I deleted all the `@method` annotations, it works doesn't matter where I put it in the common docblock.

```
/**
 * Class MembershipStatus
 * @psalm-immutable
 * @package Domain\Membership\Enums
 */
```


```
/**
 * Class MembershipStatus
 * @package Domain\Membership\Enums
 * @psalm-immutable
 */
```

**Reply by rizqyhi** [reply]
Hi @orklah I just have spare time to try test this docblock. I tried using the same docblock above in the `ImmutableAnnotationTest` test and it failed. Here is the error message:

```
Psalm\Exception\CodeException : InvalidDocblock - src/somefile.php:8:27 - UNVERIFIED_USER is not a valid method in docblock for A
```

**Reply by rizqyhi** [reply]
Ok found another interesting thing. This docblock is failing the test:

```
/**
 * ...
 * @method static MembershipStatus UNVERIFIED_USER
 * @method static MembershipStatus VERIFIED_USER
 * @psalm-immutable
 */
```

But here is fine:


```
/**
 * ...
 * @method static MembershipStatus UNVERIFIED_USER()
 * @method static MembershipStatus VERIFIED_USER()
 * @psalm-immutable
 */
```

Note the parentheses after the method name. I don't know whether this is the standard way to write the method annotation yet.

**Reply by weirdan** [reply]
Yes, `@method` names require parens.

**Reply by rizqyhi** [reply]
Thank you @weirdan, so we can confirm this wasn't a bug in Psalm. I'll just leave some references here:

- https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc-tags.md#58-method
- https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/method.html
